sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java

```diff
@@ -472,16 +475,72 @@ public <T> CosmosPagedFlux<T> queryItems(String query, Class<T> classType) {
+
+            CosmosContainerIdentity cosmosContainerIdentity = new CosmosContainerIdentity(this.database.getId(), this.id);
+            ProactiveContainerInitConfig proactiveContainerInitConfig = new ProactiveContainerInitConfigBuilder(Arrays.asList(cosmosContainerIdentity))
+                    .setProactiveConnectionRegions(1)
+                    .build();
+
+            return withContext(context -> openConnectionsAndInitCachesInternal(proactiveContainerInitConfig)
                                             .flatMap(openResult -> {
                                                 logger.info("OpenConnectionsAndInitCaches: {}", openResult);
                                                 return Mono.empty();
                                             }));
         } else {
-            logger.warn(
-                    String.format(
-                        "OpenConnectionsAndInitCaches is already called once on Container %s, no operation will take place in this call",
-                        this.getId()));
+            logger.warn("OpenConnectionsAndInitCaches is already called once on Container {}, no operation will take place in this call", this.getId());
+            return Mono.empty();
+        }
+    }
+
+    /**
+     *  Best effort to initialize the container by warming up the caches and connections to a specified no. of regions from the
+     *  preferred list of regions.
+     *
+     *  Depending on how many partitions the container has, the total time needed will also change. But generally you can use the following formula
+     *  to get an estimated time:
+     *  If it took 200ms to establish a connection, and you have 100 partitions in your container
+     *  then it will take around (100 * 4 / CPUCores) * 200ms to open all connections after get the address list
+     *
+     *  <p>
+     *  <br>NOTE: This API ideally should be called only once during application initialization before any workload.
+     *  <br>In case of any transient error, caller should consume the error and continue the regular workload.
+     *  </p>
+     * <p>
+     * In order to minimize latencies associated with warming up caches and opening connections
+     * the no. of proactive connection regions cannot be more than {@link ProactiveContainerInitConfigBuilder#MAX_NO_OF_PROACTIVE_CONNECTION_REGIONS}.
+     * </p>
+     *
+     * @param numProactiveConnectionRegions the no of regions to proactively connect to
+     * @return Mono of Void.
+     */
+    public Mono<Void> openConnectionsAndInitCaches(int numProactiveConnectionRegions) {
+
+        List<String> preferredRegions = clientAccessor.getPreferredRegions(this.database.getClient());
+        boolean endpointDiscoveryEnabled = clientAccessor.isEndpointDiscoveryEnabled(this.database.getClient());
+
+        checkArgument(numProactiveConnectionRegions > 0, "no. of proactive connection regions should be greater than 0");
+
+        if (numProactiveConnectionRegions > 1) {
```

Comment:
What if the numProactiveConnectionRegions is higher than the list of preferred regions? Ignore (and may be log a warning) or throw exception?

Reply:
^^ never mind, there is a check for that in ProactiveContainerInitConfigBuilder.

Reply:
It's a valid point, the `checkArgument` call checks for that. Semantics for `checkArgument(expression, message)` - if the expression evaluates to true - then the exception is not thrown. So the exception is only thrown if `preferredRegions != null && preferredRegions.size() >= numProactiveRegions` evaluates to `false`.